### PR TITLE
Add gem packaging

### PR DESCRIPTION
**Pull request by atongen**
Hi @xavierholt, cool project!

This PR keeps all the original logic, but adds the ability to package the code as a gem, and also adds a few tests. Let me know what you think. Thanks!

You should consider publishing this to RubyGems.org.

### CODE REVIEW

File: bin/twemproxy_exporter [hunk 1]
```diff
@@ -5,6 +5,10 @@ require 'prometheus/client/rack/exporter'
 require 'rack'
 require 'yaml'
 
+lib = File.expand_path('../../lib', __FILE__)
+$LOAD_PATH.unshift(lib) if File.directory?(lib) && !$LOAD_PATH.include?(lib)
```
Comment (xavierholt) [inline bin/twemproxy_exporter]: Is there a reason (backwards compatibility?) to prefer this to `require_relative`?  I've seen it in several projects, but I'm still a bit nervous about mucking with `$LOAD_PATH`...

File: lib/twemproxy_exporter/counter.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+module TwemproxyExporter
+  class Counter
+    def initialize(registry, name, desc)
+      @counter = Prometheus::Client::Counter.new(name, desc)
+      registry.register(@counter)
+      @last = 0
+    end
+
+    def count(value, labels = {})
```
Comment (xavierholt) [inline lib/twemproxy_exporter/counter.rb:9]: This function is my biggest open question.  The logic here is to allow the underlying twemproxy counter to be reset (like if it restarts) and keep counting up.  I'm still not sure if this is the correct response.

Curious if you have any thoughts on this.  Do you know how Prometheus handles it if an exporter reboots?

File: spec/twemproxy_exporter/gauge_spec.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+require "spec_helper"
+
+describe TwemproxyExporter::Gauge do
+  before(:all) do
+    registry = Prometheus::Client.registry
+    @gauge = TwemproxyExporter::Gauge.new(registry, :test_value, "A test gauge")
+  end
+
+  context "#count" do
+    it "should only increase" do
```
Comment (xavierholt) [inline spec/twemproxy_exporter/gauge_spec.rb]: This one should decrease too...

File: twemproxy_exporter.gemspec [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# coding: utf-8
+lib = File.expand_path('../lib', __FILE__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+require 'twemproxy_exporter/version'
+
+Gem::Specification.new do |spec|
+  spec.name          = "twemproxy_exporter"
+  spec.version       = TwemproxyExporter::VERSION
+  spec.authors       = ["xavierholt"]
+  spec.email         = ["xavierholt@example.com"]
```
Comment (xavierholt) [inline twemproxy_exporter.gemspec]: I'm @gmail.com  Go ahead and credit yourself too if you want!

**Review by xavierholt — COMMENTED** [review]
This is fantastic - thanks a ton!

This is an old project from about a year ago now, when I was both managing sharded Redis clusters and experimenting with Prometheus.  I hadn't looked at it in quite some time - glad you stumbled across it and found it useful!

I haven't published any gems before, but can certainly give it a shot.  I'm also interested in handing it off to https://github.com/prometheus if it might be generally useful.  The mainline exporters seem to be mostly Go, though...

Left a few comments in the PR code.  Nothing blocking - just a couple notes / questions.  Thanks again!

File: bin/twemproxy_exporter (hunk 1 above)
Comment (atongen) [reply bin/twemproxy_exporter]: This code is similar to what you'd see in the gemspec when you run `bundle gem new_gem`, and it pretty common to find in gems. That said, I just tested this and it looks like it's unnecessary anyway.

File: lib/twemproxy_exporter/counter.rb (hunk 2 above)
Comment (atongen) [reply lib/twemproxy_exporter/counter.rb:9]: The prometheus server should correctly handle counter resets. I think we can simplify things by dropping the counter and gauge wrapper classes and just using the prometheus ruby gem counter/gauge classes directly. A bit of more info here:
* https://prometheus.io/docs/querying/functions/#increase
* https://www.robustperception.io/how-does-a-prometheus-counter-work/

File: twemproxy_exporter.gemspec (hunk 4 above)
Comment (atongen) [reply twemproxy_exporter.gemspec]: Thanks, will do!

**Comment by atongen** [comment]
@xavierholt, thanks for the feedback. I'll push some changes shortly to reflect these comments.

File: lib/twemproxy_exporter/counter.rb (hunk 2 above)
Comment (atongen) [reply lib/twemproxy_exporter/counter.rb:9]: Actually, I looked at the data we're collecting and this seems to be working fine as-is. I think I'll keep an eye on this for a little longer before recommending any changes.

File: spec/twemproxy_exporter/gauge_spec.rb (hunk 3 above)
Comment (atongen) [reply spec/twemproxy_exporter/gauge_spec.rb]: Will update the test.

File: lib/twemproxy_exporter/counter.rb (hunk 2 above)
Comment (xavierholt) [reply lib/twemproxy_exporter/counter.rb:9]: Yeah - this is a little weird.  Prometheus itself seems to handle things as I'd expect, but the client libraries don't ([Ruby example](https://github.com/prometheus/client_ruby/blob/master/lib/prometheus/client/counter.rb#L13))...  I think the current logic is what's needed to handle a Twemproxy counter reset the same way Prometheus would.

**Comment by xavierholt** [comment]
@atongen - Looking good!  I think we misunderstood each other about the requires, though; the executable isn't currently working for me as-is (still passes tests, but I think RSpec does its own path tweaking).  Made you [a PR](https://github.com/atongen/twemproxy_exporter/pull/1) with what I had in mind.

**Comment by atongen** [comment]
Looks good, thanks @xavierholt !

**Comment by xavierholt** [comment]
Thanks @atongen!  I'll look into publishing the gem soon - maybe this afternoon if I get time.

**Pull request merged**